truss-chains/truss_chains/framework.py

```diff
@@ -40,6 +43,112 @@
 _ENTRYPOINT_ATTR_NAME = "_chains_entrypoint"
 
 ChainletT = TypeVar("ChainletT", bound=definitions.ABCChainlet)
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+# Error Collector ######################################################################
+
+
+class _ErrorKind(str, enum.Enum):
+    TYPE_ERROR = enum.auto()
+    IO_TYPE_ERROR = enum.auto()
+    MISSING_API_ERROR = enum.auto()
+
+
+class _ErrorLocation(definitions.SafeModel):
+    src_path: str
+    line: Optional[int] = None
+    chainlet_name: Optional[str] = None
+    method_name: Optional[str] = None
+
+    def __str__(self) -> str:
+        value = f"{self.src_path}:{self.line}"
+        if self.chainlet_name and self.method_name:
+            value = f"{value} ({self.chainlet_name}.{self.method_name})"
+        elif self.chainlet_name:
+            value = f"{value} ({self.chainlet_name})"
+        else:
+            assert not self.chainlet_name
+        return value
+
+
+class _ValidationError(definitions.SafeModel):
+    msg: str
+    kind: _ErrorKind
+    location: _ErrorLocation
+
+    def __str__(self) -> str:
+        return f"{self.location} [kind: {self.kind.name}]: {self.msg}"
+
+
+class _ErrorCollector:
+    _errors: list[_ValidationError]
+
+    def __init__(self) -> None:
+        self._errors = []
+        # This hook is for the case of just running the Chainlet file, without
+        # making a push - we want to surface the errors at exit.
+        atexit.register(self.maybe_display_errors)
```

Comment:
for my own curiosity -- how does this show up? Does this just print out at program exit?

Reply:
It is written to stderr: https://github.com/basetenlabs/truss/pull/1223/files#diff-99038349289851d98e0fcd8c8c53858ee1030e9ec108125ee2db5c25a786ec88R116